ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiSnapshotWorldState.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Hyperledger Besu Contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+package org.hyperledger.besu.ethereum.bonsai;
+
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.core.MutableWorldState;
+import org.hyperledger.besu.ethereum.core.SnapshotMutableWorldState;
+import org.hyperledger.besu.plugin.services.storage.SnappableKeyValueStorage;
+import org.hyperledger.besu.plugin.services.storage.SnappedKeyValueStorage;
+
+/**
+ * This class takes a snapshot of the worldstate as the basis of a mutable worldstate. It is able to
+ * commit/perist as a trielog layer only. This is useful for async blockchain opperations like block
+ * creation and/or point-in-time queries since the snapshot worldstate is fully isolated from the
+ * main BonsaiPersistedWorldState.
+ */
+public class BonsaiSnapshotWorldState extends BonsaiInMemoryWorldState
+    implements SnapshotMutableWorldState {
+
+  private final SnappedKeyValueStorage accountSnap;
+  private final SnappedKeyValueStorage codeSnap;
+  private final SnappedKeyValueStorage storageSnap;
+  private final SnappedKeyValueStorage trieBranchSnap;
+
+  private BonsaiSnapshotWorldState(
+      final BonsaiWorldStateArchive archive,
+      final BonsaiSnapshotWorldStateKeyValueStorage snapshotWorldStateStorage) {
+    super(archive, snapshotWorldStateStorage);
+    this.accountSnap = (SnappedKeyValueStorage) snapshotWorldStateStorage.accountStorage;
+    this.codeSnap = (SnappedKeyValueStorage) snapshotWorldStateStorage.codeStorage;
+    this.storageSnap = (SnappedKeyValueStorage) snapshotWorldStateStorage.storageStorage;
+    this.trieBranchSnap = (SnappedKeyValueStorage) snapshotWorldStateStorage.trieBranchStorage;
+  }
+
+  public static BonsaiSnapshotWorldState create(
```

Comment:
Is there a potential race with the storages being changed while the snapshots are generated?

Reply:
There is some risk of that theoretically, but considering we write to rocksdb in batch commits and the snapshots are essentially atomic, this is probably very very minimal chance in practice.  There is certainly far less concurrency risk than our current strategy where the underlying persisted state can change underneath a BonsaiInMemoryWorldState or BonsaiLayeredWorldState

Reply:
It seems reasonable to implement locking in BonsaiPersistedWorldState so that we do not commit a batch write (via calculateRootHash) and take a snapshot concurrently.  It should be a very short lock for getting a snapshot, and I delaying getting a snapshot while waiting on a lock would normally not be significant.

I will add this in one of the subsequent snapshot PRs and add tests :+1